kingdom/access/authorization/verify.py

```diff
@@ -0,0 +1,181 @@
+    # Sanity check.
+    assert access_request.operation & (
+        Permission.CREATE | Permission.UPDATE | Permission.DELETE
+    )
+
+    resource = access_request.resource
+    if resource not in owned_policies:
+        # Resource is unknown to subject.
+        return False
+    # resource_owned = owned_policies[resource]
+
+    # CREATE case is different than UPDATE and DELETE.
+    # It requires a "*" selector.
+    if access_request.operation == Permission.CREATE.value:
+        # To CREATE, subject must have at least one '*' policy.
+        if TOKEN_ALL not in owned_policies[resource]:
+            return False
+
+        # Checking is a simple O(1) step.
+        return is_allowed(
+            owned_policies[resource][TOKEN_ALL], access_request.operation
+        )
+        # return Permission.CREATE in owned_policies[resource][TOKEN_ALL]
+
+    # Deal with UPDATE or DELETE
+    # Check for generics.
+    if TOKEN_ALL in owned_policies[resource]:
+        permissions = owned_policies[resource][TOKEN_ALL]
+        if is_allowed(permissions, access_request.operation):
+            # Wildcard matches permission.
+            return True
+
+    # Then specify.
+    if access_request.selector not in owned_policies[resource]:
+        # No rule for requested instance. Denied.
+        return False
+
+    permissions = owned_policies[resource][access_request.selector]
+    return is_allowed(permissions, access_request.operation)
+
+
+def is_allowed(
+    owned_permissions: Union[Tuple[Permission, ...], int],
+    requested_operation: Union[int, Permission],
+) -> bool:
+    """
+    When a subject tries to perform a write, this function calculates whether
+    the requested operation is allowed on a set of owned permissions.
+
+    There are a few caveats:
+     1. Write permission overrides read permission.
+        e.g. If a Subject has any write permission, him/her implicitly
+        have read permission.
+     2. No write permissions override one-another.
+        e.g. If a Subject has DELETE permission and it doesn't have
+        UPDATE or CREATE permissions, it is only allowed to delete.
+
+
+    >>> owned_perm = (Permission.READ, Permission.UPDATE, Permission.DELETE),
+    >>> requested_op = Permission.CREATE
+    >>> is_allowed(owned_perm, requested_op)
+    False
+    """
+    # TODO: Refactor and improve typing
+
+    if isinstance(requested_operation, int):
+        requested_operation = Permission(requested_operation)
+    if isinstance(owned_permissions, int):
+        owned_permissions = (owned_permissions,)
+
+    # corner case is when requested permission is READ:
+    if requested_operation == Permission.READ and owned_permissions:
+        # if we have ANY permission, it means that the user is able to read
+        return True
+
+    permissions = 0
+    for owned_permission in owned_permissions:
+        permissions |= owned_permission
+
+    return int(permissions & requested_operation) > 0
```

Comment:
really amazing to see this conditions being performed using bit operations, this return could be just `bool(permissions & requested_operation)`

Reply:
Excellent suggestion 💪